# Final V2 Tuning on Canonical Elo Only

After the ablations, tuning is restricted to the six canonical Elo features so discarded noise cannot re-enter the model.

# כוונון סופי של V2 על Elo קנוני בלבד

מחקרי ההסרה הראו ש־DNA, מפגשים ישירים וכושר מתגלגל אינם משפרים את אות ה־Elo הקנוני בתצורה הנוכחית. לכן המחברת הסופית מסירה אותם ממטריצת האימון ומשאירה רק ארבעה דירוגים נקודתיים בזמן ושני הפרשי דירוגים. הפחתת המרחב מצמצמת הזדמנויות לפיצולים על רעש ומקלה על פירוש תוצאת הכוונון.

In [1]:
from pathlib import Path
import sys

import numpy as np
import optuna
import pandas as pd
from optuna.samplers import TPESampler
from sklearn.metrics import accuracy_score, brier_score_loss, log_loss
from xgboost import XGBClassifier

project_root = Path.cwd()
if not (project_root / "data").exists():
    project_root = project_root.parent
elo_module_dir = project_root / "src" / "features"
if str(elo_module_dir) not in sys.path:
    sys.path.insert(0, str(elo_module_dir))
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

from elo import PointInTimeEngine
from src.data.splitter import ChronologicalSplitter

RANDOM_SEED = 42

## Point-in-Time Elo and Locked Time Split

Canonical identities and deterministic chronological ordering produce pre-match features before the fixed train and validation boundaries are applied.

## בניית Elo נקודתי בזמן וחלוקת זמן נעולה

המנוע משתמש במפתחות הקבוצה הקנוניים וממיין לפי `datetime`, ולאחר מכן לפי `match_id` ו־`game_id` כדי לקבוע סדר מפות בתוך סדרה. דירוגי Global ו־Map נשמרים לפני עדכון התוצאה הנוכחית. המנוע ממשיך לחשב H2H ו־Rolling Form לשימוש עתידי, אך עמודות אלה אינן נכללות ב־allowlist של המודל.

לאחר יצירת הפיצ'רים מתבצעת החלוקה הכרונולוגית: Train עד סוף 2025, Validation ברבעון הראשון של 2026 ו־Test מאפריל 2026. רק Train ו־Validation עוברים סימטריזציה; Test נשאר נעול.

In [2]:
source_columns = [
    "match_id", "game_id", "team1_id", "team1",
    "team2_id", "team2", "is_total", "bestOf",
    "score1_game", "score2_game", "map_name", "datetime",
    "team1_win", "team1_join_key", "team2_join_key",
]
source_df = pd.read_csv(
    project_root / "data" / "final_tournament_features.csv",
    usecols=source_columns, low_memory=False,
)
point_in_time_df = PointInTimeEngine(k_factor=24, initial_rating=1500).transform(source_df)
point_in_time_df["map_position"] = (
    point_in_time_df.groupby("match_id", sort=False).cumcount() + 1
)
splits = ChronologicalSplitter().split(point_in_time_df)
train_base = splits["train"]
val_base = splits["val"]
locked_test_rows = len(splits["test"])

print(f"Train גולמי: {len(train_base):,}")
print(f"Validation גולמי: {len(val_base):,}")
print(f"Test נעול: {locked_test_rows:,}")
print(f"מפה 1 ב־Validation: {val_base['map_position'].eq(1).sum():,}")
print(f"מפות 2+ ב־Validation: {val_base['map_position'].gt(1).sum():,}")

Train גולמי: 5,472
Validation גולמי: 633
Test נעול: 595
מפה 1 ב־Validation: 274
מפות 2+ ב־Validation: 359


## Symmetrization and Minimal Allowlist

Teams are mirrored only after splitting, and the matrix contains four absolute ratings plus global and map Elo differences.

## סימטריזציה ו־allowlist מצומצם

לכל מפה נוצר עותק שבו הקבוצות ודירוגיהן מוחלפים והתווית מתהפכת. ההפרשים מחושבים לאחר ההחלפה, ובדיקה אנטי־סימטרית מחייבת היפוך סימן מדויק. הפעולה מתבצעת באופן עצמאי בכל חלוקה ורק לאחר הפיצול בזמן.

מטריצת האימון כוללת בדיוק שישה פיצ'רים: דירוג Global ודירוג Map לכל צד, והפרש קבוצה 1 פחות קבוצה 2 לכל אחד משני סוגי הדירוג.

In [3]:
rating_pairs = [
    ("team1_elo_global", "team2_elo_global"),
    ("team1_elo_map", "team2_elo_map"),
]
identity_pairs = [
    ("team1_id", "team2_id"),
    ("team1", "team2"),
    ("team1_join_key", "team2_join_key"),
]
absolute_columns = [column for pair in rating_pairs for column in pair]
feature_columns = absolute_columns + ["elo_global_diff", "elo_map_diff"]
context_columns = [
    "match_id", "game_id", "datetime", "map_name", "map_position",
    "team1_id", "team1", "team2_id", "team2",
    "team1_join_key", "team2_join_key", "team1_win",
]

def symmetrize_elo(split_df):
    original = split_df[context_columns + absolute_columns].copy().reset_index(drop=True)
    mirrored = original.copy()
    for team1_column, team2_column in identity_pairs + rating_pairs:
        mirrored[team1_column] = original[team2_column].to_numpy(copy=True)
        mirrored[team2_column] = original[team1_column].to_numpy(copy=True)
    mirrored["team1_win"] = 1 - original["team1_win"].to_numpy()

    symmetric = pd.concat([original, mirrored], ignore_index=True)
    symmetric["elo_global_diff"] = (
        symmetric["team1_elo_global"] - symmetric["team2_elo_global"]
    )
    symmetric["elo_map_diff"] = (
        symmetric["team1_elo_map"] - symmetric["team2_elo_map"]
    )

    midpoint = len(original)
    for diff_column in ["elo_global_diff", "elo_map_diff"]:
        original_values = symmetric.iloc[:midpoint][diff_column].to_numpy()
        mirrored_values = symmetric.iloc[midpoint:][diff_column].to_numpy()
        if not np.allclose(original_values, -mirrored_values):
            raise AssertionError(f"Diff sign did not flip: {diff_column}")
    if not np.isclose(symmetric["team1_win"].mean(), 0.5):
        raise AssertionError("Symmetrization did not balance the target.")
    return symmetric

train_df = symmetrize_elo(train_base)
val_df = symmetrize_elo(val_base)
X_train = train_df[feature_columns]
y_train = train_df["team1_win"].astype(int)
X_val = val_df[feature_columns]
y_val = val_df["team1_win"].astype(int)

print(f"פיצ'רים מאושרים: {feature_columns}")
print(f"Train מסומטר: {len(train_df):,}")
print(f"Validation מסומטר: {len(val_df):,}")

פיצ'רים מאושרים: ['team1_elo_global', 'team2_elo_global', 'team1_elo_map', 'team2_elo_map', 'elo_global_diff', 'elo_map_diff']
Train מסומטר: 10,944
Validation מסומטר: 1,266


## Reproducing the Fixed Baseline

The untuned Elo-only model is refit first so the tuned candidate is compared against a locally reproduced baseline.

## שחזור קו הבסיס הקבוע

לפני הכוונון מאומן מחדש מודל Elo-only בתצורה הקבועה של מחקרי ההסרה. התחזיות שלו נשמרות כדי לאפשר Bootstrap מזווג מול המודל המכוונן על אותן מפות ואותם אשכולות משחק. המדדים המתועדים משמשים גם בדיקת שחזור מול הערכים 0.608 דיוק, 0.582 דיוק במפה 1, 0.236 ברייר כללי ו־0.239 ברייר במפה 1.

In [4]:
fixed_model = XGBClassifier(
    objective="binary:logistic", eval_metric="logloss",
    n_estimators=1000, learning_rate=0.03, max_depth=3,
    min_child_weight=5, subsample=0.8, colsample_bytree=0.9,
    reg_lambda=2.0, tree_method="hist", random_state=RANDOM_SEED,
    n_jobs=-1, early_stopping_rounds=40,
)
fixed_model.fit(
    X_train, y_train,
    eval_set=[(X_train, y_train), (X_val, y_val)],
    verbose=False,
)
fixed_probability = fixed_model.predict_proba(X_val)[:, 1]
fixed_prediction = (fixed_probability >= 0.5).astype(int)
map1_mask = val_df["map_position"].eq(1)
fixed_accuracy = accuracy_score(y_val, fixed_prediction)
fixed_brier = brier_score_loss(y_val, fixed_probability)
fixed_map1_accuracy = accuracy_score(
    y_val.loc[map1_mask], fixed_prediction[map1_mask.to_numpy()]
)
fixed_map1_brier = brier_score_loss(
    y_val.loc[map1_mask], fixed_probability[map1_mask.to_numpy()]
)

print(f"דיוק קו בסיס: {fixed_accuracy:.6f}")
print(f"ברייר קו בסיס: {fixed_brier:.6f}")
print(f"דיוק קו בסיס במפה 1: {fixed_map1_accuracy:.6f}")
print(f"ברייר קו בסיס במפה 1: {fixed_map1_brier:.6f}")

דיוק קו בסיס: 0.608215
ברייר קו בסיס: 0.236055
דיוק קו בסיס במפה 1: 0.582117
ברייר קו בסיס במפה 1: 0.238658


## TPE Optimization for Validation Log-Loss

Fifty seeded Optuna trials search depth, learning rate, sampling, and regularization, with validation log-loss as the objective.

## כוונון TPE לפי לוג־לוס

Optuna מריץ 50 ניסויים עם דוגם TPE בעל seed קבוע. כל ניסוי בוחר עומק, קצב למידה, שיעורי דגימת שורות ועמודות, וענישות L1/L2. קצב הלמידה ושני פרמטרי הענישה נדגמים בסולם לוגריתמי מפני שהבדלים יחסיים חשובים יותר מהבדלים מוחלטים בטווחים אלה.

פונקציית המטרה ממזערת לוג־לוס על Validation. בכל ניסוי עצירה מוקדמת קובעת כמה עצים נדרשים בפועל. Test אינו משתתף באימון, בעצירה, בבחירת פרמטרים או בדיווח.

In [5]:
def build_tuned_model(parameters):
    return XGBClassifier(
        objective="binary:logistic", eval_metric="logloss",
        n_estimators=1200, tree_method="hist",
        random_state=RANDOM_SEED, n_jobs=-1, early_stopping_rounds=50,
        **parameters,
    )

def objective(trial):
    parameters = {
        "max_depth": trial.suggest_int("max_depth", 3, 8),
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3, log=True),
        "subsample": trial.suggest_float("subsample", 0.6, 1.0),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.6, 1.0),
        "reg_alpha": trial.suggest_float("reg_alpha", 1e-8, 1.0, log=True),
        "reg_lambda": trial.suggest_float("reg_lambda", 1e-8, 1.0, log=True),
    }
    model = build_tuned_model(parameters)
    model.fit(
        X_train, y_train,
        eval_set=[(X_train, y_train), (X_val, y_val)],
        verbose=False,
    )
    probability = model.predict_proba(X_val)[:, 1]
    trial.set_user_attr("best_iteration", int(model.best_iteration))
    return log_loss(y_val, probability, labels=[0, 1])

study = optuna.create_study(
    direction="minimize",
    sampler=TPESampler(seed=RANDOM_SEED, n_startup_trials=10),
)
study.optimize(objective, n_trials=50, gc_after_trial=True, show_progress_bar=False)

print(f"מספר ניסויים שהושלמו: {len(study.trials)}")
print(f"לוג־לוס מיטבי בחיפוש: {study.best_value:.6f}")
print("פרמטרים נבחרים:")
for parameter_name, parameter_value in study.best_params.items():
    print(f"  {parameter_name}: {parameter_value}")

[I 2026-09-15 09:44:44,100] A new study created in memory with name: no-name-6f656b56-5f70-4896-9edd-728a409af556


[I 2026-09-15 09:44:48,872] Trial 0 finished with value: 0.6664543747901917 and parameters: {'max_depth': 5, 'learning_rate': 0.2536999076681772, 'subsample': 0.892797576724562, 'colsample_bytree': 0.8394633936788146, 'reg_alpha': 1.77071686435378e-07, 'reg_lambda': 1.7699302940633311e-07}. Best is trial 0 with value: 0.6664543747901917.


[I 2026-09-15 09:45:04,017] Trial 1 finished with value: 0.6633148193359375 and parameters: {'max_depth': 3, 'learning_rate': 0.19030368381735815, 'subsample': 0.8404460046972835, 'colsample_bytree': 0.8832290311184181, 'reg_alpha': 1.4610865886287176e-08, 'reg_lambda': 0.574485163632042}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:25,828] Trial 2 finished with value: 0.6665210723876953 and parameters: {'max_depth': 7, 'learning_rate': 0.020589728197687916, 'subsample': 0.6727299868828402, 'colsample_bytree': 0.6733618039413735, 'reg_alpha': 2.716051144654844e-06, 'reg_lambda': 0.00015777981883364995}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:31,058] Trial 3 finished with value: 0.6655552387237549 and parameters: {'max_depth': 5, 'learning_rate': 0.02692655251486473, 'subsample': 0.8447411578889518, 'colsample_bytree': 0.6557975442608167, 'reg_alpha': 2.1734877073417355e-06, 'reg_lambda': 8.528933855762793e-06}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:32,951] Trial 4 finished with value: 0.6642276048660278 and parameters: {'max_depth': 5, 'learning_rate': 0.14447746112718687, 'subsample': 0.6798695128633439, 'colsample_bytree': 0.8056937753654446, 'reg_alpha': 0.0005486767416600901, 'reg_lambda': 2.3528990899815284e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:36,888] Trial 5 finished with value: 0.6656783223152161 and parameters: {'max_depth': 6, 'learning_rate': 0.0178601378893971, 'subsample': 0.6260206371941118, 'colsample_bytree': 0.9795542149013333, 'reg_alpha': 0.530953226900921, 'reg_lambda': 0.02932100047183291}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:44,453] Trial 6 finished with value: 0.6646318435668945 and parameters: {'max_depth': 4, 'learning_rate': 0.013940346079873234, 'subsample': 0.8736932106048627, 'colsample_bytree': 0.7760609974958406, 'reg_alpha': 9.469038421774442e-08, 'reg_lambda': 9.149877525022172e-05}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:46,484] Trial 7 finished with value: 0.6643173694610596 and parameters: {'max_depth': 3, 'learning_rate': 0.22038218939289875, 'subsample': 0.7035119926400067, 'colsample_bytree': 0.8650089137415928, 'reg_alpha': 3.116654126398047e-06, 'reg_lambda': 0.00014472520367197597}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:51,303] Trial 8 finished with value: 0.6654974818229675 and parameters: {'max_depth': 6, 'learning_rate': 0.01875220945578641, 'subsample': 0.9878338511058234, 'colsample_bytree': 0.9100531293444458, 'reg_alpha': 0.32808889626606236, 'reg_lambda': 0.14408501080722544}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:54,586] Trial 9 finished with value: 0.6691113710403442 and parameters: {'max_depth': 6, 'learning_rate': 0.22999586428143728, 'subsample': 0.6353970008207678, 'colsample_bytree': 0.6783931449676581, 'reg_alpha': 2.300479202014574e-08, 'reg_lambda': 4.005370050283172e-06}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:57,025] Trial 10 finished with value: 0.6650298237800598 and parameters: {'max_depth': 4, 'learning_rate': 0.08984346146296351, 'subsample': 0.9614408650708839, 'colsample_bytree': 0.8570692283464406, 'reg_alpha': 1.5180704314836276e-06, 'reg_lambda': 0.05560003965887478}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:45:59,484] Trial 11 finished with value: 0.6634024977684021 and parameters: {'max_depth': 4, 'learning_rate': 0.11500343465131521, 'subsample': 0.7484902859154928, 'colsample_bytree': 0.9637492421746525, 'reg_alpha': 7.17963913520317e-06, 'reg_lambda': 8.742249644124113e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:02,486] Trial 12 finished with value: 0.6641631126403809 and parameters: {'max_depth': 4, 'learning_rate': 0.06427290891400102, 'subsample': 0.7458218708277714, 'colsample_bytree': 0.909590112172339, 'reg_alpha': 5.273235100733669e-08, 'reg_lambda': 2.464621765409409e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:04,020] Trial 13 finished with value: 0.6639917492866516 and parameters: {'max_depth': 3, 'learning_rate': 0.1258339089094697, 'subsample': 0.7052084670705713, 'colsample_bytree': 0.9626172516623243, 'reg_alpha': 3.794082139377309e-07, 'reg_lambda': 0.06501848684701646}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:05,635] Trial 14 finished with value: 0.6645314693450928 and parameters: {'max_depth': 3, 'learning_rate': 0.27049790831917975, 'subsample': 0.8560223683693958, 'colsample_bytree': 0.817412748064673, 'reg_alpha': 1.7124033595940522e-07, 'reg_lambda': 0.0004259734985716855}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:07,486] Trial 15 finished with value: 0.6648136973381042 and parameters: {'max_depth': 4, 'learning_rate': 0.07943032034947517, 'subsample': 0.7947948378066236, 'colsample_bytree': 0.84353382872851, 'reg_alpha': 2.6599424330318942e-08, 'reg_lambda': 0.004921829432106071}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:09,751] Trial 16 finished with value: 0.6645193696022034 and parameters: {'max_depth': 3, 'learning_rate': 0.11320508853350174, 'subsample': 0.8657441051405743, 'colsample_bytree': 0.9767452008059001, 'reg_alpha': 4.056865861412559e-06, 'reg_lambda': 0.10266477569475915}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:11,701] Trial 17 finished with value: 0.6634371876716614 and parameters: {'max_depth': 3, 'learning_rate': 0.2841888943758047, 'subsample': 0.9816937937365989, 'colsample_bytree': 0.9111958798870186, 'reg_alpha': 2.0415412147819965e-08, 'reg_lambda': 8.657025892024346e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:13,822] Trial 18 finished with value: 0.663834273815155 and parameters: {'max_depth': 3, 'learning_rate': 0.05293504685567811, 'subsample': 0.7752767142402552, 'colsample_bytree': 0.962802907266154, 'reg_alpha': 4.154351872332084e-07, 'reg_lambda': 7.155049553751509e-05}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:15,154] Trial 19 finished with value: 0.6637465357780457 and parameters: {'max_depth': 3, 'learning_rate': 0.23985746560145393, 'subsample': 0.9155635693861152, 'colsample_bytree': 0.8975899075499648, 'reg_alpha': 1.6298397934941175e-08, 'reg_lambda': 0.02365002484897168}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:16,674] Trial 20 finished with value: 0.6643131971359253 and parameters: {'max_depth': 3, 'learning_rate': 0.19706339563415695, 'subsample': 0.7950479954257299, 'colsample_bytree': 0.9115050360876309, 'reg_alpha': 8.926320823168299e-07, 'reg_lambda': 1.8146920260542975e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:18,055] Trial 21 finished with value: 0.6679054498672485 and parameters: {'max_depth': 4, 'learning_rate': 0.26822403006137896, 'subsample': 0.9564832448261353, 'colsample_bytree': 0.8477281251365375, 'reg_alpha': 2.090121296117854e-05, 'reg_lambda': 2.861078857178331e-07}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:19,711] Trial 22 finished with value: 0.6648662686347961 and parameters: {'max_depth': 4, 'learning_rate': 0.24364677203689533, 'subsample': 0.9683044214675441, 'colsample_bytree': 0.9272925537441513, 'reg_alpha': 3.8091721823147996e-08, 'reg_lambda': 3.545459330645735e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:21,023] Trial 23 finished with value: 0.6654148101806641 and parameters: {'max_depth': 4, 'learning_rate': 0.24879752823844323, 'subsample': 0.8879797738037913, 'colsample_bytree': 0.9575119017875453, 'reg_alpha': 5.737075674517432e-07, 'reg_lambda': 3.082326045299758e-07}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:23,590] Trial 24 finished with value: 0.6644247770309448 and parameters: {'max_depth': 3, 'learning_rate': 0.15290767326534924, 'subsample': 0.9155768516700821, 'colsample_bytree': 0.8610674496468661, 'reg_alpha': 1.5923978659399754e-08, 'reg_lambda': 2.271382182300722e-05}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:24,918] Trial 25 finished with value: 0.6657074689865112 and parameters: {'max_depth': 4, 'learning_rate': 0.20341793744290626, 'subsample': 0.8353696032743356, 'colsample_bytree': 0.9505167057025337, 'reg_alpha': 1.0359658418870744e-08, 'reg_lambda': 5.66448815232519e-06}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:32,577] Trial 26 finished with value: 0.6648929119110107 and parameters: {'max_depth': 3, 'learning_rate': 0.19175238802019892, 'subsample': 0.7459563476124936, 'colsample_bytree': 0.8192446938716424, 'reg_alpha': 2.592786726571131e-08, 'reg_lambda': 2.19247070086676e-06}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:34,131] Trial 27 finished with value: 0.665493905544281 and parameters: {'max_depth': 4, 'learning_rate': 0.29291838832108025, 'subsample': 0.7740161231723235, 'colsample_bytree': 0.8601754626112141, 'reg_alpha': 1.1346605009395695e-08, 'reg_lambda': 0.14116337870532797}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:39,263] Trial 28 finished with value: 0.6639971733093262 and parameters: {'max_depth': 4, 'learning_rate': 0.16504176487833241, 'subsample': 0.9666493150137397, 'colsample_bytree': 0.8487262270726861, 'reg_alpha': 2.103199268302206e-07, 'reg_lambda': 1.537329677350536e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:40,785] Trial 29 finished with value: 0.6653252243995667 and parameters: {'max_depth': 4, 'learning_rate': 0.14602019571545172, 'subsample': 0.8717287644001461, 'colsample_bytree': 0.8860002256838165, 'reg_alpha': 4.44029138750771e-07, 'reg_lambda': 9.405632483099577e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:42,650] Trial 30 finished with value: 0.6646623015403748 and parameters: {'max_depth': 3, 'learning_rate': 0.11617127489639466, 'subsample': 0.9870240074750571, 'colsample_bytree': 0.9576934756553578, 'reg_alpha': 2.455731743480422e-08, 'reg_lambda': 2.4440329843344385e-08}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:44,285] Trial 31 finished with value: 0.6653517484664917 and parameters: {'max_depth': 3, 'learning_rate': 0.22020007748726245, 'subsample': 0.7726682558543039, 'colsample_bytree': 0.8666670386863152, 'reg_alpha': 8.573140853992547e-08, 'reg_lambda': 0.6640186627220734}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:45,768] Trial 32 finished with value: 0.6650975346565247 and parameters: {'max_depth': 3, 'learning_rate': 0.15044922635494598, 'subsample': 0.8724065638876899, 'colsample_bytree': 0.8354575480917263, 'reg_alpha': 6.561367909059349e-08, 'reg_lambda': 0.04723659175140247}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:47,500] Trial 33 finished with value: 0.6650310158729553 and parameters: {'max_depth': 3, 'learning_rate': 0.2774412020720508, 'subsample': 0.9357307316339307, 'colsample_bytree': 0.8113869164574494, 'reg_alpha': 1.4696400996322784e-07, 'reg_lambda': 0.32234946893645383}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:48,858] Trial 34 finished with value: 0.6667603254318237 and parameters: {'max_depth': 4, 'learning_rate': 0.1952161339670412, 'subsample': 0.8936458194235234, 'colsample_bytree': 0.9365921781250941, 'reg_alpha': 1.6727049059475835e-08, 'reg_lambda': 0.1326001324197097}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:50,515] Trial 35 finished with value: 0.664991557598114 and parameters: {'max_depth': 3, 'learning_rate': 0.13392833119816464, 'subsample': 0.9154172130670063, 'colsample_bytree': 0.8744430543801274, 'reg_alpha': 1.407259790686356e-08, 'reg_lambda': 0.4192820783001727}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:51,982] Trial 36 finished with value: 0.6642310619354248 and parameters: {'max_depth': 3, 'learning_rate': 0.14986917409373618, 'subsample': 0.9573698054476594, 'colsample_bytree': 0.9019556265354106, 'reg_alpha': 1.826538030651134e-07, 'reg_lambda': 1.8693397311630263e-07}. Best is trial 1 with value: 0.6633148193359375.


[I 2026-09-15 09:46:53,850] Trial 37 finished with value: 0.6633034944534302 and parameters: {'max_depth': 3, 'learning_rate': 0.24343481689010896, 'subsample': 0.9344118207704185, 'colsample_bytree': 0.9710474671226088, 'reg_alpha': 3.824855907161912e-06, 'reg_lambda': 4.318644427085766e-07}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:46:55,926] Trial 38 finished with value: 0.6654102206230164 and parameters: {'max_depth': 3, 'learning_rate': 0.20210007977767008, 'subsample': 0.8813278368725708, 'colsample_bytree': 0.8428039324140699, 'reg_alpha': 1.673014225747526e-08, 'reg_lambda': 2.9365920518768317e-08}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:46:57,729] Trial 39 finished with value: 0.6636230945587158 and parameters: {'max_depth': 3, 'learning_rate': 0.20056534204245835, 'subsample': 0.9237812969697223, 'colsample_bytree': 0.8961303773848813, 'reg_alpha': 1.3228571341868067e-07, 'reg_lambda': 1.105489837718519e-05}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:46:59,254] Trial 40 finished with value: 0.6638510227203369 and parameters: {'max_depth': 3, 'learning_rate': 0.18889957296297538, 'subsample': 0.9692936549282064, 'colsample_bytree': 0.9410679326845003, 'reg_alpha': 7.452233030246552e-06, 'reg_lambda': 2.439239085320239e-08}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:47:01,940] Trial 41 finished with value: 0.6641213893890381 and parameters: {'max_depth': 3, 'learning_rate': 0.2944282655674877, 'subsample': 0.9638503781954395, 'colsample_bytree': 0.9915657397491868, 'reg_alpha': 1.8002428776869083e-06, 'reg_lambda': 7.192555214337611e-08}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:47:03,724] Trial 42 finished with value: 0.6638253927230835 and parameters: {'max_depth': 3, 'learning_rate': 0.18771927305677308, 'subsample': 0.9693596562678773, 'colsample_bytree': 0.9163621868299139, 'reg_alpha': 4.851412321713639e-08, 'reg_lambda': 1.2771156287614235e-06}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:47:05,330] Trial 43 finished with value: 0.6649196743965149 and parameters: {'max_depth': 3, 'learning_rate': 0.19477661212538003, 'subsample': 0.8778330952642098, 'colsample_bytree': 0.8740773521810251, 'reg_alpha': 1.277046172578662e-07, 'reg_lambda': 2.1429029917884917e-07}. Best is trial 37 with value: 0.6633034944534302.


[I 2026-09-15 09:47:06,857] Trial 44 finished with value: 0.6630033850669861 and parameters: {'max_depth': 3, 'learning_rate': 0.22473870152166872, 'subsample': 0.9744377938554797, 'colsample_bytree': 0.9620639280745505, 'reg_alpha': 1.4445201932091805e-06, 'reg_lambda': 1.6603528779515085e-05}. Best is trial 44 with value: 0.6630033850669861.


[I 2026-09-15 09:47:08,441] Trial 45 finished with value: 0.6636595726013184 and parameters: {'max_depth': 4, 'learning_rate': 0.22995139638774112, 'subsample': 0.9416044649726598, 'colsample_bytree': 0.9165909747384238, 'reg_alpha': 4.062572990745495e-06, 'reg_lambda': 4.557968762192747e-07}. Best is trial 44 with value: 0.6630033850669861.


[I 2026-09-15 09:47:10,246] Trial 46 finished with value: 0.6640492081642151 and parameters: {'max_depth': 3, 'learning_rate': 0.2953389008717481, 'subsample': 0.9252810835957178, 'colsample_bytree': 0.9434685437704348, 'reg_alpha': 2.2847538122422806e-08, 'reg_lambda': 2.1519465317257883e-08}. Best is trial 44 with value: 0.6630033850669861.


[I 2026-09-15 09:47:12,646] Trial 47 finished with value: 0.664476752281189 and parameters: {'max_depth': 3, 'learning_rate': 0.20615289991235897, 'subsample': 0.965305367886234, 'colsample_bytree': 0.8820887065916242, 'reg_alpha': 4.2326241516311835e-06, 'reg_lambda': 2.3676693520653953e-08}. Best is trial 44 with value: 0.6630033850669861.


[I 2026-09-15 09:47:14,068] Trial 48 finished with value: 0.6643006205558777 and parameters: {'max_depth': 4, 'learning_rate': 0.2773235099148248, 'subsample': 0.9392086258302176, 'colsample_bytree': 0.9085193206342885, 'reg_alpha': 4.691509111261365e-07, 'reg_lambda': 1.772393814708473e-08}. Best is trial 44 with value: 0.6630033850669861.


[I 2026-09-15 09:47:16,155] Trial 49 finished with value: 0.6641484498977661 and parameters: {'max_depth': 3, 'learning_rate': 0.21560061711680267, 'subsample': 0.9608098034926944, 'colsample_bytree': 0.9654341192493555, 'reg_alpha': 7.100173644959831e-08, 'reg_lambda': 3.6921734883734123e-07}. Best is trial 44 with value: 0.6630033850669861.


מספר ניסויים שהושלמו: 50
לוג־לוס מיטבי בחיפוש: 0.663003
פרמטרים נבחרים:
  max_depth: 3
  learning_rate: 0.22473870152166872
  subsample: 0.9744377938554797
  colsample_bytree: 0.9620639280745505
  reg_alpha: 1.4445201932091805e-06
  reg_lambda: 1.6603528779515085e-05


## Refit and Validation Stratification

The selected configuration is refit on training data and evaluated overall and by map position.

## אימון התצורה הנבחרת ופילוח Validation

המודל הנבחר מאומן מחדש על Train עם עצירה מוקדמת מול Validation. מדווחים לוג־לוס, דיוק וברייר בכלל החלוקה, ולאחר מכן דיוק וברייר במפה 1 ובמפות 2 ומעלה. מיקום המפה משמש רק לפילוח ואינו פיצ'ר.

In [6]:
best_model = build_tuned_model(study.best_params)
best_model.fit(
    X_train, y_train,
    eval_set=[(X_train, y_train), (X_val, y_val)],
    verbose=10,
)
tuned_probability = best_model.predict_proba(X_val)[:, 1]
tuned_prediction = (tuned_probability >= 0.5).astype(int)

def calculate_tuned_metrics(mask):
    selected_target = y_val.loc[mask]
    selected_probability = tuned_probability[mask.to_numpy()]
    selected_prediction = tuned_prediction[mask.to_numpy()]
    return {
        "rows": int(mask.sum()),
        "accuracy": accuracy_score(selected_target, selected_prediction),
        "brier": brier_score_loss(selected_target, selected_probability),
    }

metric_groups = {
    "כלל האימות": pd.Series(True, index=val_df.index),
    "מפה 1": val_df["map_position"].eq(1),
    "מפה 2 ומעלה": val_df["map_position"].gt(1),
}
tuned_metrics = {
    group_name: calculate_tuned_metrics(mask)
    for group_name, mask in metric_groups.items()
}
final_log_loss = log_loss(y_val, tuned_probability, labels=[0, 1])
print(f"האיטרציה הטובה ביותר: {best_model.best_iteration}")
print(f"לוג־לוס סופי: {final_log_loss:.6f}")
for group_name, metrics in tuned_metrics.items():
    print(
        f"{group_name}: שורות={metrics['rows']:,}, "
        f"דיוק={metrics['accuracy']:.6f}, ברייר={metrics['brier']:.6f}"
    )

[0]	validation_0-logloss:0.68277	validation_1-logloss:0.67973


[10]	validation_0-logloss:0.66032	validation_1-logloss:0.66583


[20]	validation_0-logloss:0.65295	validation_1-logloss:0.66986


[30]	validation_0-logloss:0.64805	validation_1-logloss:0.67127


[40]	validation_0-logloss:0.64319	validation_1-logloss:0.67507


[50]	validation_0-logloss:0.63877	validation_1-logloss:0.67887


[57]	validation_0-logloss:0.63583	validation_1-logloss:0.67922


האיטרציה הטובה ביותר: 7
לוג־לוס סופי: 0.663003
כלל האימות: שורות=1,266, דיוק=0.610585, ברייר=0.235418
מפה 1: שורות=548, דיוק=0.589416, ברייר=0.238312
מפה 2 ומעלה: שורות=718, דיוק=0.626741, ברייר=0.233209


## Paired Cluster-Bootstrap Guardrail

Both models predict the same rows, so matches are resampled jointly and paired metric differences preserve within-match dependence.

## Guardrail סטטיסטי באתחול מחדש מזווג

שני המודלים מייצרים תחזית לאותן שורות, ולכן בכל חזרת Bootstrap נדגמים 274 אשכולות `match_id` עם החזרה וכל השורות של המשחק נשמרות יחד. השוואה מזווגת זו מבטלת חלק מהשונות המשותפת ומודדת ישירות את השינוי שנוצר מהכוונון.

פער הדיוק מוגדר כדיוק המודל המכוונן פחות דיוק המודל הקבוע. שיפור הברייר מוגדר כברייר המודל הקבוע פחות ברייר המודל המכוונן. ערך חיובי מעדיף את הכוונון. החישוב נעשה הן בכלל Validation והן במפה 1, שהיא שכבת ה־pre-series המחמירה.

In [7]:
bootstrap_iterations = 1000
bootstrap_rng = np.random.default_rng(RANDOM_SEED)
target_values = y_val.to_numpy()
map1_values = map1_mask.to_numpy()
tuned_correct = (tuned_prediction == target_values).astype(float)
fixed_correct = (fixed_prediction == target_values).astype(float)
tuned_squared_error = (tuned_probability - target_values) ** 2
fixed_squared_error = (fixed_probability - target_values) ** 2

bootstrap_rows = pd.DataFrame({
    "match_id": val_df["match_id"].to_numpy(),
    "rows": 1.0,
    "tuned_correct": tuned_correct,
    "fixed_correct": fixed_correct,
    "tuned_brier_sum": tuned_squared_error,
    "fixed_brier_sum": fixed_squared_error,
    "map1_rows": map1_values.astype(float),
    "map1_tuned_correct": tuned_correct * map1_values,
    "map1_fixed_correct": fixed_correct * map1_values,
    "map1_tuned_brier_sum": tuned_squared_error * map1_values,
    "map1_fixed_brier_sum": fixed_squared_error * map1_values,
})
cluster_metrics = bootstrap_rows.groupby("match_id", sort=False).sum()
cluster_values = cluster_metrics.to_numpy(dtype=float)
match_count = len(cluster_values)
sample_indices = bootstrap_rng.integers(
    0, match_count, size=(bootstrap_iterations, match_count)
)
sample_totals = cluster_values[sample_indices].sum(axis=1)
rows = sample_totals[:, 0]
map1_rows = sample_totals[:, 5]

tuned_accuracy_samples = sample_totals[:, 1] / rows
tuned_brier_samples = sample_totals[:, 3] / rows
overall_accuracy_delta = (sample_totals[:, 1] - sample_totals[:, 2]) / rows
overall_brier_improvement = (sample_totals[:, 4] - sample_totals[:, 3]) / rows
map1_tuned_accuracy_samples = sample_totals[:, 6] / map1_rows
map1_tuned_brier_samples = sample_totals[:, 8] / map1_rows
map1_accuracy_delta = (sample_totals[:, 6] - sample_totals[:, 7]) / map1_rows
map1_brier_improvement = (sample_totals[:, 9] - sample_totals[:, 8]) / map1_rows

def confidence_interval(values):
    return np.quantile(values, [0.025, 0.975])

overall_accuracy_ci = confidence_interval(tuned_accuracy_samples)
overall_brier_ci = confidence_interval(tuned_brier_samples)
overall_accuracy_delta_ci = confidence_interval(overall_accuracy_delta)
overall_brier_improvement_ci = confidence_interval(overall_brier_improvement)
map1_accuracy_ci = confidence_interval(map1_tuned_accuracy_samples)
map1_brier_ci = confidence_interval(map1_tuned_brier_samples)
map1_accuracy_delta_ci = confidence_interval(map1_accuracy_delta)
map1_brier_improvement_ci = confidence_interval(map1_brier_improvement)

print(f"מספר משחקים ייחודיים: {match_count}; חזרות Bootstrap: {bootstrap_iterations}")
print(f"דיוק מכוונן — רווח סמך 95%: [{overall_accuracy_ci[0]:.6f}, {overall_accuracy_ci[1]:.6f}]")
print(f"ברייר מכוונן — רווח סמך 95%: [{overall_brier_ci[0]:.6f}, {overall_brier_ci[1]:.6f}]")
print(f"שינוי דיוק מול קו הבסיס — רווח סמך 95%: [{overall_accuracy_delta_ci[0]:.6f}, {overall_accuracy_delta_ci[1]:.6f}]")
print(f"שיפור ברייר מול קו הבסיס — רווח סמך 95%: [{overall_brier_improvement_ci[0]:.6f}, {overall_brier_improvement_ci[1]:.6f}]")
print(f"דיוק מכוונן במפה 1 — רווח סמך 95%: [{map1_accuracy_ci[0]:.6f}, {map1_accuracy_ci[1]:.6f}]")
print(f"ברייר מכוונן במפה 1 — רווח סמך 95%: [{map1_brier_ci[0]:.6f}, {map1_brier_ci[1]:.6f}]")
print(f"שינוי דיוק במפה 1 — רווח סמך 95%: [{map1_accuracy_delta_ci[0]:.6f}, {map1_accuracy_delta_ci[1]:.6f}]")
print(f"שיפור ברייר במפה 1 — רווח סמך 95%: [{map1_brier_improvement_ci[0]:.6f}, {map1_brier_improvement_ci[1]:.6f}]")

מספר משחקים ייחודיים: 274; חזרות Bootstrap: 1000
דיוק מכוונן — רווח סמך 95%: [0.572556, 0.649594]
ברייר מכוונן — רווח סמך 95%: [0.225775, 0.245452]
שינוי דיוק מול קו הבסיס — רווח סמך 95%: [-0.006280, 0.011528]
שיפור ברייר מול קו הבסיס — רווח סמך 95%: [-0.000499, 0.001800]
דיוק מכוונן במפה 1 — רווח סמך 95%: [0.534672, 0.653285]
ברייר מכוונן במפה 1 — רווח סמך 95%: [0.224706, 0.250949]
שינוי דיוק במפה 1 — רווח סמך 95%: [-0.005474, 0.023723]
שיפור ברייר במפה 1 — רווח סמך 95%: [-0.001302, 0.001992]


## Tuning Decision

Because the tuned model does not statistically clear the fixed baseline, production retains the simpler default canonical-Elo configuration.

## מסקנת הכוונון ונקודת החלטה

המודל המכוונן השיג לוג־לוס 0.663003, דיוק 0.610585 וברייר 0.235418 בכלל Validation. במפה 1 התקבלו דיוק 0.589416 וברייר 0.238312; במפות 2+ התקבלו דיוק 0.626741 וברייר 0.233209. לעומת התצורה הקבועה, השיפור הנקודתי הוא 0.002370 בדיוק ו־0.000637 בברייר בכלל החלוקה, ו־0.007299 בדיוק ו־0.000346 בברייר במפה 1.

רווח הסמך המזווג לשינוי הדיוק הכולל הוא ‎[-0.006280, 0.011528]‎ ורווח הסמך לשיפור הברייר הכולל הוא ‎[-0.000499, 0.001800]‎. במפה 1 הרווחים הם ‎[-0.005474, 0.023723]‎ לדיוק ו־‎[-0.001302, 0.001992]‎ לברייר. כל ארבעת הרווחים כוללים אפס, ולכן הכוונון אינו מנקה סטטיסטית את קו הבסיס הקבוע במדגם זה. התצורה הנבחרת נשמרת כמועמדת ניסויית בלבד; Test נשאר ללא סימטריזציה, ללא חיזוי וללא הערכה.